modules/sops/templates/default.nix

```diff
@@ -0,0 +1,108 @@
+    (replaceStrings [ "@subst@" ] [ "${subst-pairs}" ] (readFile ./subs.py));
+  subst-pairs = pkgs.writeText "pairs" (concatMapStringsSep "\n" (name:
+    "${toString config.sops.placeholder.${name}} ${
+      config.sops.secrets.${name}.path
+    }") (attrNames config.sops.secrets));
+  coercibleToString = mkOptionType {
+    name = "coercibleToString";
+    description = "value that can be coerced to string";
+    check = strings.isCoercibleToString;
+    merge = mergeEqualOption;
+  };
+  templateType = submodule ({ config, ... }: {
+    options = {
+      name = mkOption {
+        type = str;
+        default = config._module.args.name;
+        description = ''
+          Name of the file used in /run/secrets-rendered
+        '';
+      };
+      path = mkOption {
+        type = str;
+        default = "/run/secrets-rendered/${config.name}";
+      };
+      content = mkOption {
+        type = lines;
+        default = "";
+        description = ''
+          Content of the file
+        '';
+      };
+      mode = mkOption {
+        type = str;
+        default = "0400";
+        description = ''
+          Permissions mode of the rendered secret file in octal.
+        '';
+      };
+      owner = mkOption {
+        type = str;
+        default = "root";
+        description = ''
+          User of the file.
+        '';
+      };
+      group = mkOption {
+        type = str;
+        default = users.${config.owner}.group;
+        description = ''
+          Group of the file.
+        '';
+      };
+      file = mkOption {
+        type = types.path;
+        default = pkgs.writeText config.name config.content;
+        visible = false;
+        readOnly = true;
+      };
+    };
+  });
+in {
+  options.sops = {
+    templates = mkOption {
+      type = attrsOf templateType;
+      default = { };
+    };
+    placeholder = mkOption {
+      type = attrsOf coercibleToString;
+      default = { };
+      visible = false;
+    };
+    substituteCmd = mkOption {
+      type = types.path;
+      default = substitute;
+    };
+  };
+
+  config = optionalAttrs (options ? sops.secrets)
+    (mkIf (config.sops.templates != { }) {
+      sops.placeholder = mapAttrs
+        (name: _: mkDefault "<SOPS:${hashString "sha256" name}:PLACEHOLDER>")
+        config.sops.secrets;
+
+      system.activationScripts.renderSecrets = mkIf (cfg.templates != { })
+        (stringAfter ([ "setupSecrets" ]
+          ++ optional (secretsForUsers != { }) "setupSecretsForUsers") ''
+            echo Setting up sops templates...
+            ${concatMapStringsSep "\n" (name:
+              let tpl = config.sops.templates.${name};
+              in ''
+                mkdir -p "${dirOf tpl.path}"
+                ${config.sops.substituteCmd} ${tpl.file} > ${tpl.path}
+                chmod "${tpl.mode}" "${tpl.path}"
+                chown "${tpl.owner}" "${tpl.path}"
+                chgrp "${tpl.group}" "${tpl.path}"
```

Comment:
That might leak the secret for small amount of time when permissions are not applied corretly yet. Better to use the `install` command here to make this step atomically in one command.

Reply:
install -D -m "${tpl.mode}" -o "${tpl.owner}" -g "${tpl.group}" "${tpl.file}" "${tpl.path}"

Reply:
Mhm, since this is using `stdout` this is not so easy to replace with the install command. Maybe the substitution command should also install the secret? Otherwise we also need to use `(umask 077; <yourcommand> > secret)` to make sure that no other process can read the secret when it's written.

Reply:
Is this flow secure enough for creating a secret file:
1. umask 077
2. write the file
3. set group
4. set owner
5. set mode

Reply:
What if touch the file and set owner & permission before writing to it?

Reply:
Touch and set perms afterwards is racy. Someone could open the file before the perms are restricted.

Reply:
That's why `umask` would first make the file not readable by any non-root process.